src/App.js

```diff
@@ -710,6 +711,70 @@ function FullApp() {
     [chainId, active]
   );
 
+  const { data: currentRewardRound, error: failedFetchingRoundRewards } = useSWR(
+    [getServerUrl(chainId, "/tradingRewards"), "latest"],
+    {
+      fetcher: (url, round) => fetch(`${url}&round=${round}`).then((res) => res.json()),
+    }
+  );
+
+  const [leaderboardData, setLeaderboardData] = useState([]);
+
+  // Get volume, position and reward from user round data
+  const userPosition = useMemo(() => {
+    if (!leaderboardData) {
+      return undefined;
+    }
+    const leaderBoardIndex = leaderboardData?.findIndex(
+      (trader) => trader.user_address.toLowerCase() === account?.toLowerCase()
+    );
+
+    return leaderBoardIndex + 1;
+  }, [account, leaderboardData]);
+
+  useEffect(() => {
+    if (!currentRewardRound || !!currentRewardRound?.message) {
+      return undefined;
+    }
+    const traders = currentRewardRound.rewards
+      ?.sort((a, b) => b.volume - a.volume)
+      .map((trader) => {
+        const positionReward = ethers.BigNumber.from(trader.reward);
+        const degenReward = ethers.BigNumber.from(trader.degen_reward);
+
+        return {
+          ...trader,
+          totalReward: positionReward.add(degenReward),
+          positionReward,
+          degenReward,
+        };
+      }); // Sort traders by highest to lowest in volume
+    setLeaderboardData(traders);
+  }, [currentRewardRound]);
+
+  const updateLeaderboardOptimistically = (volumeToAdd) => {
+    const _leaderboardData = [...leaderboardData];
+    const currentUser = _leaderboardData.find(
+      (trader) => trader?.user_address?.toLowerCase() === account?.toLowerCase()
+    );
+    // If user exists, update volume
+    if (currentUser) {
+      currentUser.volume = ethers.BigNumber.from(currentUser.volume).add(ethers.BigNumber.from(volumeToAdd).toString());
+    }
+    // Otherwise insert at the end of the list after users without volume
+    else {
+      const firstUserIndexWithoutVolume = leaderboardData.findIndex((trader) => trader?.volume === "0");
+      const userItem = {
+        user_address: account,
+        volume: volumeToAdd,
+        reward: 0,
+        degen_reward: 0,
+      };
+      _leaderboardData.splice(firstUserIndexWithoutVolume, 0, userItem);
+    }
+    setLeaderboardData(_leaderboardData);
+  };
+
```

Comment:
I think we could seperate this out into a context to make it a bit more maintanable

Reply:
also should re-order with added volume on